src/protocol/modulus_conversion/convert_shares.rs

```diff
@@ -286,7 +289,7 @@ mod tests {
                     let tweaked = tweak.flip_bit(ctx.role(), triple);
 
                     let v = MaliciousValidator::new(ctx);
-                    let m_ctx = v.context();
+                    let m_ctx = v.context().set_total_records(1).set_total_upgrades(1);
```

Comment:
Seeing this reminds me that the root malicious context held by the validator can't have a `set_total_records(self, ...)` method because we need something that modifies the context in place in that case.  I don't quite know what to do about that right now.

Is it possible to pass this information in to the validator so that it constructs that root context correctly instead?  WDYT?

@akoshelev and I continue to go back and forth about the value of the split between validator and malicious context; perhaps we need to have another look at that.

Reply:
I will look into this.

Reply:
It looks like this mostly works, but with one notable difficulty. The test `protocol::basics::mul::sparse::test::check_output_malicious` relies on narrowing the upgrade context at runtime (via `upgrade_with`). Processing the inputs as records 0/1 of a common step doesn't work, I'm assuming because the zero positions are different for the two inputs so the channel usage is not consistent.

It seems a little weird to me to fix an upgrade count at the time the validator is created, but allow dynamic narrowing of the upgrade context at runtime.

The impl of `IntoMalicious` for tuples may also be problematic. If there are cases where the two sides of the tuple differ in length and differ in upgrade mechanism, that would be a problem (but I think that case may not currently exist). And the way I see to support "either same length or same mechanism" requires behaving differently depending on which of those two cases applies, which is also a bit gross.

Reply:
https://github.com/andyleiserson/ipa/commit/3aba756ecc4b50de208f62c40191cf611bcdd5c7 has my experiments